### PR TITLE
KEYCLOAK-17936 FAPI-CIBA : support Signed Authentication Request

### PR DESCRIPTION
**Pull request by tnorimat**
This PR is for [KEYCLOAK-11846 OpenID Connect Financial-grade API: Client Initiated Backchannel Authentication Profile](https://issues.redhat.com/browse/KEYCLOAK-11846), also is the part of the project [FAPI-CIBA(poll mode)](https://github.com/keycloak/kc-sig-fapi/projects) of [FAPI-SIG](https://github.com/keycloak/kc-sig-fapi) activity.

Based on the merged [OIDC CIBA](https://github.com/keycloak/keycloak/pull/7679) feature, this PR provides additional feature to satisfy [FAPI-CIBA(poll mode) security profile](https://bitbucket.org/openid/fapi/src/master/Financial_API_WD_CIBA.md). [KEYCLOAK-17936 FAPI-CIBA : support Signed Authentication Request](https://issues.redhat.com/browse/KEYCLOAK-17936) describes it more.

### CODE REVIEW

**Comment by tnorimat** [comment]
@mposolda This PR is for supporting FAPI-CIBA security profile support based on the merged OIDC CIBA feature as mentioned in [the 17th FAPI-SIG meeting](https://github.com/keycloak/kc-sig-fapi/blob/master/FAPI-SIG/meetings/17th/presentations/FAPI-SIG_17th_MTG_agenda.pdf). Could you or your team review it?

**Review by mposolda — CHANGES_REQUESTED** [review]
@tnorimat Thanks for the PR!

I've added one inline comment regarding signature algorithms. Also I have few other comments:

- It will be nice that ClientPolicyExecutor `SecureRequestObjectExecutor` is triggered somehow for the CIBA Authentication  request. Similarly `SecureSigningAlgorithmExecutor` (or other executor) will be probably nice to be triggered to makes sure that for the FAPI, the secured algorithms are used. But I am not sure whether to integrate with client policies here or in separate PR. Maybe it makes sense to rather use separate PR for this? As I suppose it will require to have the CIBA Authorization request event available as the type of `ClientPolicyEvent`, which is not yet available.

**Comment by tnorimat** [comment]
@mposolda 

As you said in https://github.com/keycloak/keycloak/pull/8006#pullrequestreview-691877966 , I've been preparing the PR extending client policies to treat requests on Backchannel Authentication Endpoint.

Also, I've been preparing the PR extending some existing executors to meet the requirements by [FAPI-CIBA](https://bitbucket.org/openid/fapi/src/master/Financial_API_WD_CIBA.md) .

Therefore, I would like to work as follows.

1. Revising this PR.
2. After (1)'s PR is merged, implementing PR for extending client policies to treat requests on Backchannel Authentication Endpoint based on (1)'s PR.
3. After (2)'s PR is merged, implementing PR for extending some existing executors to meet the requirements by FAPI-CIBA based on (1) and (2)'s PR 

WDYT?

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/OIDCWellKnownProviderTest.java [hunk 1]
```diff
@@ -173,6 +173,7 @@ public void testDiscovery() {
             assertEquals(oidcConfig.getBackchannelAuthenticationEndpoint(), oauth.getBackchannelAuthenticationUrl());
             assertContains(oidcConfig.getGrantTypesSupported(), OAuth2Constants.CIBA_GRANT_TYPE);
             Assert.assertNames(oidcConfig.getBackchannelTokenDeliveryModesSupported(), "poll");
+            Assert.assertNames(oidcConfig.getBackchannelAuthenticationRequestSigningAlgValuesSupported(), "none", Algorithm.PS256, Algorithm.PS384, Algorithm.PS512, Algorithm.RS256, Algorithm.RS384, Algorithm.RS512, Algorithm.ES256, Algorithm.ES384, Algorithm.ES512, Algorithm.HS256, Algorithm.HS384, Algorithm.HS512);
```
[the remark this answers is not in the document]
Comment (tnorimat) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/OIDCWellKnownProviderTest.java:176]: @mposolda Yes, that's right.
I would like to revise this PR as follows.
- `none` is not allowed. keycloak returns an error if CD(client) sends backchannel authentication requests with not signed authentication request.
- ES*, PS* and RS* are allowed. keycloak returns an error if CD(client) sends backchannel authentication requests with signed authentication request by HS*.
- add arquillian integration tests to check the things just mentioned above.

By the way, [FAPI-CIBA](https://bitbucket.org/openid/fapi/src/master/Financial_API_WD_CIBA.md#markdown-header-710-algorithm-considerations) does not allow RS*. I would like to implement the PR for extending some existing executors to meet the requirements by FAPI-CIBA.

**Comment by tnorimat** [comment]
@mposolda
Aside from it, how about treating requests on Device Authorization Endpoint? If you agree with it, I would like to implement PR for it.

File: testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/OIDCWellKnownProviderTest.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (mposolda) [reply testsuite/integration-arquillian/tests/base/src/test/java/org/keycloak/testsuite/oidc/OIDCWellKnownProviderTest.java:176]: +1 to everything you mentioned in the last paragraph. We can allow RS* by default, but reject it with the client policy executor similarly like we're doing for other use-cases where RS* is rejected by the executor.

**Comment by mposolda** [comment]
+1 to the stages you proposed here. Also +1 for the Device Authorization Endpoint

**Comment by mposolda** [comment]
@tnorimat I've opened the pull request https://github.com/Hitachi/keycloak/pull/3 to update this pull request. I've added method "isAsymmetricAlgorithm" on SignatureProvider class to be able to filter algorithms programatically instead of having hardcoded filtering of concrete provider. Hardcoded filtering may not reliably work for example for the case if someone adds new symmetric algorithm providers in the future etc. If you are fine with this, feel free to merge my PR to your branch (which should effectively update this PR against Keycloak master)

**Comment by tnorimat** [comment]
@mposolda Thank you. I've merged your PR into this PR. Could you check it?

**Review by mposolda — APPROVED** [review]
@tnorimat Thanks!

**Comment by mposolda** [comment]
@tnorimat FYI. I've created one minor follow-up JIRA https://issues.redhat.com/browse/KEYCLOAK-18583 . Figured that there is this unused method, which I forgot to remove in my PR...

**Comment by tnorimat** [comment]
@mposolda I'll work on https://issues.redhat.com/browse/KEYCLOAK-18583 as soon as possible.

**Comment by mposolda** [comment]
@tnorimat Thanks! FYI. I've also created https://issues.redhat.com/browse/KEYCLOAK-18587 . Please let me know what do you think about it and if my understanding of the specification and description of the JIRA is correct.

**Comment by tnorimat** [comment]
@mposolda I would like to check and work on https://issues.redhat.com/browse/KEYCLOAK-18587 and https://issues.redhat.com/browse/KEYCLOAK-18588 .

**Pull request merged**